source/extensions/filters/common/rbac/config.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+
+#include "envoy/config/filter/http/rbac/v2/rbac.pb.h"
+#include "envoy/config/filter/network/rbac/v2/rbac.pb.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "extensions/filters/common/rbac/engine_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RBAC {
+
+/**
+ * All stats for the RBAC filter. @see stats_macros.h
+ */
+// clang-format off
+#define ALL_RBAC_FILTER_STATS(COUNTER)                                                             \
+  COUNTER(allowed)                                                                                 \
+  COUNTER(denied)                                                                                  \
+  COUNTER(shadow_allowed)                                                                          \
+  COUNTER(shadow_denied)
+// clang-format on
+
+/**
+ * Wrapper struct for RBAC filter stats. @see stats_macros.h
+ */
+struct RoleBasedAccessControlFilterStats {
+  ALL_RBAC_FILTER_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+enum class EnforcementMode { Enforced, Shadow };
+
+template <class ConfigType>
+absl::optional<RoleBasedAccessControlEngineImpl> createEngine(const ConfigType& config,
+                                                              bool disable_http_rules = false) {
+  return config.has_rules() ? absl::make_optional<RoleBasedAccessControlEngineImpl>(
+                                  config.rules(), disable_http_rules)
+                            : absl::nullopt;
+}
+
+template <class ConfigType>
+absl::optional<RoleBasedAccessControlEngineImpl>
+createShadowEngine(const ConfigType& config, bool disable_http_rules = false) {
+  return config.has_shadow_rules() ? absl::make_optional<RoleBasedAccessControlEngineImpl>(
+                                         config.shadow_rules(), disable_http_rules)
+                                   : absl::nullopt;
+}
+
+class RoleBasedAccessControlRouteSpecificFilterConfig : public Router::RouteSpecificFilterConfig {
+public:
+  RoleBasedAccessControlRouteSpecificFilterConfig(
+      const envoy::config::filter::http::rbac::v2::RBACPerRoute& per_route_config);
+
+  const absl::optional<RoleBasedAccessControlEngineImpl>& engine(EnforcementMode mode) const {
+    return mode == EnforcementMode::Enforced ? engine_ : shadow_engine_;
+  }
+
+private:
+  const absl::optional<RoleBasedAccessControlEngineImpl> engine_;
+  const absl::optional<RoleBasedAccessControlEngineImpl> shadow_engine_;
+};
+
+/**
+ * Configuration for the RBAC filter.
+ */
+class RoleBasedAccessControlFilterConfig {
```

Comment:
Still not convinced we should combine HTTP RBAC and TCP RBAC config in same class. You can have template functions and stats macros above in common/rbac/config_utility. having a filter_name in `engine()` below seems hacky.

Reply:
Okay, Moved back to http and network filter respectively, renamed `config` to `utility`.